metrics/tags.go

```diff
@@ -156,6 +156,82 @@ var _ interface {
 	json.Unmarshaler
 } = &TagSet{}
 
+// TagsAndMeta is a helper type that provides easy group manipulation of the
+// indexed Tags and the non-indexed Metadata values together. While both of them
+// are part of a metric Sample, the TagsAndMeta type isn't used there because
+// the Tags participate in the Sample's TimeSeries and the Metadata does not.
+//
+// Instead, this type is mostly used by the VU and JS code while it assembles
+// the final consolidated Tags and Metadata values to put in the Sample.
+//
+// IMPORTANT: this data structure is not thread safe! The methods of the
+// lib.VUStateTags type are where the synchronization happens.
+type TagsAndMeta struct {
+	Tags     *TagSet
+	Metadata map[string]string // could be nil
+}
+
+// SetTag adds the given key=value tag to the Tags and removes it from the
+// Metadata, if it was present there.
+func (tm *TagsAndMeta) SetTag(key, value string) {
+	tm.Tags = tm.Tags.With(key, value)
+	if tm.Metadata != nil {
+		delete(tm.Metadata, key)
+	}
+}
+
+// SetMetadata adds the given key=value datum to the Metadata and removes it
+// from the Tags, if it was present there.
+func (tm *TagsAndMeta) SetMetadata(key, value string) {
+	tm.Tags = tm.Tags.Without(key)
+	if tm.Metadata == nil {
+		tm.Metadata = map[string]string{key: value}
+	} else {
+		tm.Metadata[key] = value
+	}
+}
+
+// SetSystemTagOrMetaIfEnabled checks if the supplied SystemTag is enabled in
+// this test run (i.e. is in the SystemTagSet) and passes it to
+// SetSystemTagOrMeta() if it is.
+func (tm *TagsAndMeta) SetSystemTagOrMetaIfEnabled(enabledSystemTags *SystemTagSet, tag SystemTag, value string) {
+	if !enabledSystemTags.Has(tag) {
+		return
+	}
+	tm.SetSystemTagOrMeta(tag, value)
+}
+
+// SetSystemTagOrMeta automatically adds the system tag either as an indexed tag
+// or an unindexed one (metadata), based on the NonIndexableSystemTags set.
+func (tm *TagsAndMeta) SetSystemTagOrMeta(tag SystemTag, value string) {
+	if NonIndexableSystemTags.Has(tag) {
+		tm.SetMetadata(tag.String(), value)
+	} else {
+		tm.SetTag(tag.String(), value)
+	}
+}
+
+// Delete removes the given key from both the tags and the metadata.
+func (tm *TagsAndMeta) Delete(key string) {
+	tm.Tags = tm.Tags.Without(key)
+	if tm.Metadata != nil {
+		delete(tm.Metadata, key)
+	}
+}
+
+// Clone makes a complete copy of the current TagsAndMeta object and returns it.
+func (tm TagsAndMeta) Clone() TagsAndMeta {
+	res := TagsAndMeta{Tags: tm.Tags}
```

Comment:
Just double-checking since Atlas struct has pointers; it's a safe way to cone it?

Reply:
Yes, it has pointers precisely so it can be cloned safely like this. An `*atlas.Node` pointer is essentially a way to represent a static `key: value` string map. If two `*atlas.Node` pointers with the same root are equal, then they contain the same map values.

This is why it is vitally important for all `TagSet` values in a k6 test run to always start from the same atlas root node: https://github.com/grafana/k6/blob/4ae7d1e8db25b60433c77b7fde18f458b444e168/metrics/registry.go#L23-L25